pkg/api/errors/pubsub.go

```diff
@@ -17,170 +17,229 @@ import (
+		fmt.Sprintf("failed deserializing metadata. Error: %s", err),
+		"ERR_PUBSUB_REQUEST_METADATA",
+		"METADATA_DESERIALIZATION",
+	)
+}
 
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
+func (p *PubSubMetadataError) CloudEventCreation() error {
+	return p.build(
+		codes.InvalidArgument,
 		http.StatusInternalServerError,
-		message,
+		"cannot create cloudevent",
 		"ERR_PUBSUB_CLOUD_EVENTS_SER",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixMarshalEnvelope, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+		"CLOUD_EVENT_CREATION",
+	)
 }
 
-func PubSubMarshalEvents(name string, pubsubType string, topic string, metadata map[string]string) error {
-	message := fmt.Sprintf("error marshaling events to bytes for topic %s pubsub %s", topic, name)
-
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
+func (p *PubSubTopicError) MarshalEnvelope() error {
+	msg := fmt.Sprintf(
+		"error when marshalling cloud event envelope for topic %s pubsub %s: %s",
+		p.topic, p.p.p.name, p.p.metadata["error"],
+	)
+	return p.p.build(
+		codes.InvalidArgument,
 		http.StatusBadRequest,
-		message+". error: "+metadata["error"],
+		msg,
 		"ERR_PUBSUB_EVENTS_SER",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixMarshalEvents, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+		"MARSHAL_ENVELOPE",
+	)
 }
 
-// PubSubUnMarshalEvents only occurs in http/api.go
-func PubSubUnMarshalEvents(name string, pubsubType string, topic string, metadata map[string]string, err error) error {
-	message := fmt.Sprintf("error when unmarshaling the request for topic %s pubsub %s: %s", topic, name, err.Error())
+func (p *PubSubTopicError) MarshalEvents() error {
+	message := fmt.Sprintf(
+		"error marshaling events to bytes for topic %s pubsub %s. error: %s",
+		p.topic, p.p.p.name, p.p.metadata["error"],
+	)
 
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
+	return p.p.build(
+		codes.InvalidArgument,
 		http.StatusBadRequest,
 		message,
 		"ERR_PUBSUB_EVENTS_SER",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixUnMarshalEvents, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+		"MARSHAL_EVENTS",
+	)
 }
 
-func PubSubPublishForbidden(name string, pubsubType string, topic string, appID string, err error) error {
-	message := fmt.Sprintf("topic %s is not allowed for app id %s", topic, appID)
-	return kiterrors.NewBuilder(
-		grpcCodes.PermissionDenied,
-		http.StatusForbidden,
-		message,
-		"ERR_PUBSUB_FORBIDDEN",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixForbidden, map[string]string{"topic": topic, "error": err.Error()}).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
-}
+// UnmarshalEvents only occurs in http/api.go
+func (p *PubSubTopicError) UnmarshalEvents(err error) error {
+	message := fmt.Sprintf(
+		"error when unmarshaling the request for topic %s pubsub %s: %s",
+		p.topic, p.p.p.name, err,
```

Comment:
`p.p.p.name` - is that a typo or is it really accessed in this manner now?

Reply:
Yes, it is because we are wrapping with more context which this error type needs. We can pick a different variable name if you prefer.